### PR TITLE
[Week 10] Clarify how MCTS works & cleanup notebook

### LINKED ISSUE #212: week10/seminar_mcts: rollout() should return immediate_reward if is_done
The comment in the docstring

```
Note 2: if node is terminal (self.is_done is True), just return 0
```

is inconsistent with the suggested implementation (in the seminar recording), where the `for` loop collects reward until `is_done`, inclusive.

Solution: replace 0 with `self.immediate_reward`.

### LINKED ISSUE #213: week10/seminar_mcts: discards value for expanded child
The suggested solution for the `plan_mcts` function involves a call to `node.propagate(child_reward)`. However, this way, the child from which the rollout is computed is not updated. A workaround is to call `child.propagate(child_reward - child.immediate_reward)`; however, this is a hack, and a cleaner solution should be implemented.

### PR DESCRIPTION
**Pull request by dniku**
Fixes #212 by amending the appropriate docstring. Fixes #213 by adding a comment.

### CODE REVIEW

File: week10_planning/seminar_MCTS.ipynb [hunk 1]
```diff
@@ -6,46 +6,58 @@
    "source": [
     "## Seminar: Monte-carlo tree search (5 pts)\n",
     "\n",
-    "In this seminar, we'll implement a vanilla MCTS planning and use it to solve some Gym envs.\n",
+    "Monte Carlo tree search (MCTS) is a heuristic search algorithm, which shows cool results in challenging domains such as Go and chess. The algorithm involves building a search tree, iteratively traversion  and evaluation the nodes with Monte-Carlo simulation.\n",
```
Comment (dniku) [inline week10_planning/seminar_MCTS.ipynb:9]: This does not read quite right:

> The algorithm involves building a search tree, iteratively traversion  and evaluation the nodes with Monte-Carlo simulation.

How about this?

> The algorithm builds a search tree, iteratively traverses it, and evaluates its nodes using a Monte-Carlo simulation.

File: week10_planning/seminar_MCTS.ipynb [hunk 2]
```diff
@@ -6,46 +6,58 @@
...
    "cell_type": "markdown",
    "metadata": {},
    "source": [
     "__How it works?__\n",
-    "\n",
-    "We just start with an empty tree and from there we iterate and expand it. There are several common procedures.\n",
+    "We just start with an empty tree and expand it. There are several common procedures.\n",
     "\n",
     "__1) Selection__\n",
+    "Starting from the root, recursively select the node that corresponds tree policy.  \n",
```
Comment (dniku) [inline week10_planning/seminar_MCTS.ipynb:24]: that corresponds **to the** tree policy

File: week10_planning/seminar_MCTS.ipynb [hunk 3]
```diff
@@ -6,46 +6,58 @@
...
     "\n",
-    "Starting from the root, recursively select node with highest UCB-1 score - on the one hand, the most beneficial action, on the other hand, the least explored.\n",
+    "Widely used _tree policy_ - choose the action with the highest UCB-1 score. On one hand, this is the most beneficial action; on the other hand, it's the least explored one.\n",
```
Comment (dniku) [inline week10_planning/seminar_MCTS.ipynb:26]: I don't understand what `Widely used _tree policy_` is supposed to mean here.

File: week10_planning/seminar_MCTS.ipynb [hunk 4]
```diff
@@ -6,46 +6,58 @@
...
     "where: \n",
-    "- $N$ - number of time-steps so far,\n",
-    "- $n_a$ - times action a is taken\n",
-    "- $C_n$ - balance between exploration-exploitation, for $R \\in [0,1]$ $ C_p = \\frac{1}{\\sqrt{2}}$ \n",
-    "[paper](http://citeseerx.ist.psu.edu/viewdoc/download?doi=10.1.1.374.1202&rep=rep1&type=pdf)\n",
+    "- $N$ - number of times we have visited state $s$,\n",
+    "- $n_a$ - number of times we have taken action $a$,\n",
+    "- $C_p$ - exploration balance parameter, which is performed between exploration and exploitation. \n",
     "\n",
+    "Using Hoeffding inequality for rewards $R \\in [0,1]$ it can be shown [[3]](#3) that optimal $C_p = 1/\\sqrt{2}$. For rewards outside this range, the parameter should be tuned. We'll to try use 10 but you can tune it.\n",
```
Comment (dniku) [inline week10_planning/seminar_MCTS.ipynb:45]: > We'll to try use 10 but you can tune it.

→ 

> We'll be using 10, but you can experiment with other values.

File: week10_planning/seminar_MCTS.ipynb (hunk 3 above)
Comment (neer201) [reply week10_planning/seminar_MCTS.ipynb:26]: I mean, that, there are several tree policies, which we saw earlier as exploration strategies: epsilon-greedy, Thomson sampling, UCB-1. But was shown that in MCTS good results can achieve with UCB-1. And further, we will consider the one, but you can try to use others.

File: week10_planning/seminar_MCTS.ipynb (hunk 1 above)
Comment (neer201) [reply week10_planning/seminar_MCTS.ipynb]: Your option is better. Thanks.

File: week10_planning/seminar_MCTS.ipynb (hunk 4 above)
Comment (neer201) [reply week10_planning/seminar_MCTS.ipynb]: Yep, sounds better =)

File: week10_planning/seminar_MCTS.ipynb (hunk 2 above)
Comment (neer201) [reply week10_planning/seminar_MCTS.ipynb]: Fixed

File: week10_planning/seminar_MCTS.ipynb [hunk 5]
```diff
@@ -6,46 +6,59 @@
...
-    "\n",
-    "We just start with an empty tree and from there we iterate and expand it. There are several common procedures.\n",
+    "We just start with an empty tree and expand it. There are several common procedures.\n",
     "\n",
     "__1) Selection__\n",
+    "Starting from the root, recursively select the node that corresponds to the tree policy.  \n",
     "\n",
-    "Starting from the root, recursively select node with highest UCB-1 score - on the one hand, the most beneficial action, on the other hand, the least explored.\n",
+    "There are several options for tree policies, which we saw earlier as exploration strategies: epsilon-greedy, Thomson sampling, UCB-1. It was shown that in MCTS, UCB-1 achieves a good result. Further, we will consider the one, but you can try to use others.\n",
+    "What actions are chosen with UCB-1?. On one hand, this is the most beneficial action; on the other hand, it's the least explored one.\n",
```
Comment (dniku) [inline week10_planning/seminar_MCTS.ipynb:27]: One last touch: you have a mismatch between "actions" (plural) and "this is the most beneficial..." (singular). How about this?

> Following the UCB-1 tree policy, we will choose an action that, on one hand, we expect to have the highest return, and on the other hand, we haven't explored much.

Comment (neer201) [reply week10_planning/seminar_MCTS.ipynb:27]: Thanks again. Your option is cooler. =)

**Pull request merged**